feat($compile): add support for arbitrary property and event bindings

jbedard opened the pull request:
Here's the initial work for arbitrary property and event bindings. I think `ng-attr-*` and `ng-on-*` are essentially ready. The main thing to do is the `$sce` logic which I'm very unfamiliar with (I've never used it!), so any help/feedback in that area would be great. Also docs...

Could also think about doing the `ng-bindon-*` as @gkalpak has been calling it  (something equivalent to `[(*)]` in Angular).


**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

TODO:
- [x] no or not in this pr ~~update `getTrustedAttrContext` to the `"el|attr"` style (like `getTrustedPropContext`) to reuse code?~~
- [x] no or not in this pr  ~~add ability to add new attribute contexts?~~
- [x] run custom-elements-everywhere [angularjs tests](https://github.com/webcomponents/custom-elements-everywhere/tree/master/libraries/angularjs): [done](https://github.com/jbedard/custom-elements-everywhere/tree/angularjs-event-props) but needs to reference the new release
- [x] tests to verify effects of `addPropertySecurityContext` (fae50665e)
- [x] tests for non-trusted img/audio/video[src] values (14119a9e3)
- [x] tests for non-trusted source/track[src] values (14119a9e3)
- [x] try moving tests to new files (7157abe)
- [x] change `ngProp` to do preassign in prelink (13bab70bfa3ad733fafc050e1431e32f9b7a6021)
- [x] `ng-on-*` docs (https://github.com/angular/angular.js/pull/16627)
- [x] `ng-prop-*` docs (make note of https://github.com/angular/angular.js/pull/16614#discussion_r198717403) (https://github.com/angular/angular.js/pull/16627)
- [x] tests for `*|style` (04224c79c8d36193e64e377e506c655f36ca6cbe)
- [x] ensure tests for each context type, whitelisted, non whitelisted, trusted-value (a93ee19 - mainly copied from interp tests of the attribute equivalents)
- [x] `<source ng-prop-srcset="">` tests similar to `<img ng-prop-srcset="">` (d1632437590510be3597bac87d6300875fe3e67b)
- [x] sce tests for things like innerHTML (f547b0d)
- [x] add sce valueOf similar to ngBindHtml + tests (f547b0d)
- [x] undo `source[srcset]` *attribute* change (ef075360ab71c2ea6113c039cfb21a9988cc8ea7)
- [x] drop tests copied from `ng-attr-*` that are interpolate specific (1e6b068b309d56181669fc502ed7f2b86d6e7e87)
- [x] file bug for `ng-attr-broken_camels` issue (#16624, referenced in tests 58d4c16)
- [x] ~reduce size of sce prop IIFE~ seems unnecessary, see https://github.com/angular/angular.js/pull/16614#discussion_r200537268

jbedard committed:
refactor($compile): move img[srcset] sanitizing to helper method

jbedard committed:
feat($compile): add support for arbitrary DOM property and event bindings

Properties:

Previously only arbitrary DOM attribute bindings were supported via interpolation such as
`my-attribute="{{expression}}"` or `ng-attr-my-attribute="{{expression}}"`, and only a set of
distinct properties could be bound. `ng-prop-*` adds support for binding expressions to any DOM
properties. For example `ng-prop-foo="x"` will assign the value of the expression `x` to the
`foo` property, and re-assign whenever the expression `x` changes.

Events:

Previously only a distinct set of DOM events could be bound using directives such as `ng-click`,
`ng-blur` etc. `ng-on-*` adds support for binding expressions to any DOM event. For example
`ng-on-bar="barOccured($event)"` will add a listener to the "bar" event and invoke the
`barOccured($event)` expression.

Since HTML attributes are case-insensitive, property and event names are specified in snake_case
for `ng-prop-*` and `ng-on-*`. For example, to bind property `fooBar` use `ng-prop-foo_bar`, to
listen to event `fooBar` use `ng-on-foo_bar`.

Fixes #16428
Fixes #16235
Closes #16614

gkalpak reviewed (commented):
I haven't looked at the tests, but the rest looks reasonable :tada:

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,84 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   */
+  (function setupPropertyContexts() {
+    function registerContext(ctx, values) {
+      forEach(values, function(v) { PROP_CONTEXTS[v.toLowerCase()] = ctx; });
```
I'd prefer keeping it so the properties below can have the standard property casing. This also makes the list of properties almost the exact same as https://github.com/angular/angular/blob/6.0.6/packages/compiler/src/schema/dom_security_schema.ts#L31-L58

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,84 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+      'object|codebase',  'object|data',
+      'script|src',
+    ]);
+  })();
```
I'd say since we already have attribute sanitizing built in we should probably do the same for properties?

gkalpak replied:
I wonder how much this affects the overall size. Since these are strings, they can't be minified, but there are less of them than I thought there would be and they probably compress fine. So, I don't think this is a problem, but am still curious about the size increase.

jbedard replied:
As a quick (and inaccurate) test, this setup method on it's own is 1.2kb and 0.7kb minfied. Should minify a bit better when not isolated and then gzip would help a bit more...

petebacondarwin replied:
Size is a bit of a concern, especially because there might be very limited take up of this feature.
What if we put the core changes into angular.js but always throw an exception if the `ngProp` module is not loaded. Then we could move as much of this sanitization (and anything else auxiliary) into that module and keep the core changes to a minimum?

jbedard replied:
I'd really prefer if it was usable without adding another file/module, especially since that file will just be this property list... all the actual functionality is baked into `$compile` and will already be there. Although I guess the `ng-on-*` is the one that will be used more, so that one is more important to me to be baked in.

Is it mainly just the size?

I could probably make this a bit smaller (concatenate all the strings instead of arrays of strings...).

jbedard replied:
I mentioned I'd try to make this minify better... looks like the closure compiler already does what I was thinking of (convert `['a', 'b', 'c']` to `'a b c'.split(' ')`).

When minifying only this IIFE it is 371 bytes gzipped. When minifying the full library, then measuring just this IIFE it is 348 bytes gzipped (however it might have been minified to gzip better along with the rest of the app?).

So basically ~350 bytes is a good estimate.

jbedard commented on src/ng/compile.js:
```diff
@@ -3350,9 +3451,54 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+                'ng- versions (such as ng-click or ng-on-click instead of ng-prop-onclick) instead.');
+      }
+
+      var nodeName = nodeName_(node);
```
I wanted to keep the signatures of `getTrustedPropContext` and `getTrustedAttrContext` to be the same. WDYT?

jbedard replied:
However it looks like `getTrustedAttrContext ` is also called right beside a `nodeName_(node)` call like this, so I could change both...

jbedard replied:
Done (https://github.com/jbedard/angular.js/commit/bd636890b2c5d4294cedd9e642357f99b9cfb7bd)

jbedard commented on src/ng/compile.js:
```diff
@@ -3350,9 +3451,54 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+              scope.$watch(fn, function propertyWatchActionFn(value) {
+                if (value) {
+                  // Sanitize img[srcset] + source[srcset] values.
+                  if ((nodeName === 'img' || nodeName === 'source') && propName === 'srcset') {
```
I'll try this, but will also do it for attributes since the code is the exact same. Should have a fixup commit doing this shortly...

jbedard replied:
Done, but instead of different watchActionFns I created different `sanitizer` methods that a common watchActionFn invokes...

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,84 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * @param {string} the element name or '*' to match any element
+   * @param {string} the property name
+   */
+  this.addPropertyContext = function(ctx, elementName, propertyName) {
```
I added the `ctxoverride` error for this

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,614 @@ describe('$compile', function() {
+        $rootScope.name = 'angular';
+        $rootScope.$apply();
+        log('digest=' + element.prop('myName'));
+        expect(log).toEqual('compile=undefined; preLinkP101=undefined; preLinkP0=undefined; postLink=undefined; postCompile=undefined; digest=angular');
```
Here's one I wasn't sure about...

With `ng-attr-*` the attribute is [evaluated on pre-link](https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L3397-L3400) so it is available to other link functions before the first `$watch` applies. I have not done that for `ng-prop-*`. Should we?

Narretz replied:
I don't think that's very important for properties. Listening to attributes is / was quite important in AngularJS, but with props that's different.

petebacondarwin replied:
Just make sure it is documented, so that people don't assume it works the same as ngAttr.

jbedard replied:
I think for `$attr` it is important because properties are put into `$attrs` which is part of the directive compile/link API. But there is no equivalent API for properties. Unless you consider `$element.prop/attr` the API...

I think I'd vote to make it the same as `ngAttr` just for consistency.  Unless someone can think of a case where assigning to a DOM property one extra time (init + `$watch`) would be an issue?

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,620 @@ describe('$compile', function() {
+        element = $compile('<img ng-prop-src="testUrl"></img>')($rootScope);
+        // Some browsers complain if you try to write `javascript:` into an `img[src]`
+        // So for the test use something different
+        $rootScope.testUrl = $sce.trustAsMediaUrl('someuntrustedthing:foo();');
```
Note that for the `ng-attr-` version of this test this string was `someUnstrustedThing:foo()`. But it seems like for properties the browser (at least some) change the protocol to lowercase when assigned to the property, so the `expect` below failed (which also has the mixed case for the attribute version). I changed both this line and the `expect` to be all lower case to avoid this since I don't think it matters here...

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,620 @@ describe('$compile', function() {
+        expect(element.prop('src')).toEqual('someuntrustedthing:foo();');
+      }));
+
+      it('should sanitize concatenated values even if they are trusted', inject(function($rootScope, $compile, $sce) {
```
Is this test and the next correct for expressions like this? I'm not sure if expressions vs interpolation are or should be handled different...?

jbedard replied:
Delete?

Narretz replied:
I think so. Was the consensus yesterday that ngProp can / should only consider the result of all expressions?

jbedard commented on src/ng/compile.js:
```diff
@@ -2245,43 +2337,58 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-                }
+
+            if (isNgProp) {
+              attrs[ngAttrName] = value;
```
Is this correct? Which attribute should have the `value? If any?

Narretz replied:
Is attrs the attrs object that gets passed to the link fn / controller? I don't think ngProp should set the attribute here.

gkalpak replied:
Seems right to me.

jbedard replied:
Yes this `attrs` is what gets passed to the compile / link / controller.

Note that here we are assigning `attrs.ngPropFoo`, not `attrs.foo` like `ng-attr-foo` would. Also the value is the expression that will later get `$parsed`.  While I agree it's weird putting anything prop related on `attrs`, I did this so we can...
a) use the existing `createEventDirective` from ngEventDirs.js
b) follow the same pattern as ngEventDirs.js and `ng-attr-*` where the value that gets `$parse`-ed (or `$interpolate`-ed for `ng-attr`) is read of the attribute in the compile step of the directive

I'm not sure how to achieve both of those without putting the expression onto `attrs`...

Ideas? Opinions?

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,620 @@ describe('$compile', function() {
+        });
+      });
+
+      it('should use $$sanitizeUri on concatenated trusted values', function() {
```
Delete?

Narretz commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,620 @@ describe('$compile', function() {
+        expect(element.prop('src')).toEqual('unsafe:untrusted:foo();untrusted:foo();');
+      }));
+
+      it('should not sanitize other properties', inject(function($compile, $rootScope) {
```
I feel like this is obsolete, too. It doesn't make sense in ngProp context at least.

Narretz commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,620 @@ describe('$compile', function() {
+      });
+
+      it('should sanitize all uris in srcset', inject(function($rootScope, $compile) {
+        element = $compile('<img srcset="{{testUrl}}"></img>')($rootScope);
```
Shouldn't this be `ng-prop-srcset="testUrl"`?

gkalpak commented on src/ng/compile.js:
```diff
@@ -1909,7 +2001,7 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
               return template.replace(/\{\{/g, startSymbol).replace(/}}/g, endSymbol);
         },
-        NG_ATTR_BINDING = /^ngAttr[A-Z]/;
+        NG_PREFIX_BINDING = /^ng(Attr|Prop|On)([A-Z].*)$/;
```
Nit: With this rregex it is not possible to bind to properties that don't start with a lowercase letter (e.g. `Foo`, `_bar`, `$baz`). If we decide not to support those, we should at least document the limitations.

jbedard replied:
That means `ng-attr-*` always had that problem? :thinking:

gkalpak replied:
Yes, but is it unusual for attribute to start with an uppercase letter, `_` or `$` :smiley:
(I guess it is not very likely for properties either 🤔 - at least not for public properties you would want to bind to :wink:)

jbedard replied:
I think I'd like to fix this. Basically just change the end of the regex to `(_?[A-Z].*)$`?

gkalpak replied:
If you want to support uppercase first letter, it should be `((?:_[a-z]|[A-Z]).*)$` (because in order to bind to `Foo`, you would need `ng-prop-_foo`, which would be normalized to `ngProp_foo`).

jbedard replied:
What about properties starting with an `_`? 🤔

gkalpak replied:
You could bind to them with `__foo` (which is super-weird). But `_`-prefixed properties in JS usually means "private property", which in turn means you probably shouldn't be binding to it :grin:
(BTW, there is the same issue with attributes containing an underscore. Essentially, if you want to bind to attr/prop (using `ngAttr*`/`ngProp*`) that contains a `_`, you need to "escape" it wlike this: `__` (two underscores).)

jbedard replied:
... and attributes.  Where should that be documented though?

Could also fix it though...

gkalpak commented on src/ng/compile.js:
```diff
@@ -2245,43 +2337,58 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             ngAttrName = directiveNormalize(name);
-            isNgAttr = NG_ATTR_BINDING.test(ngAttrName);
-            if (isNgAttr) {
+            if (ngPrefixMatch = ngAttrName.match(NG_PREFIX_BINDING)) {
```
Shouldn't there be an extra pair of parenthesis here, to indicate the assignment is intentional?

gkalpak commented on src/ng/compile.js:
```diff
@@ -3350,9 +3456,56 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
 
     function addAttrInterpolateDirective(node, directives, value, name, isNgAttr) {
-      var trustedContext = getTrustedContext(node, name);
+      var tag = nodeName_(node);
```
Why not `nodeName` for consistency with `addPropertyDirective` :grin:

jbedard replied:
I think because `getTrustedAttrContext` used `tag`, but I agree and will update this one to `nodeName`...

gkalpak commented on src/ng/compile.js:
```diff
@@ -3350,9 +3456,56 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+          return {
+            pre: function ngPropPreLinkFn(scope, $element) {
+              scope.$watch(fn, function propertyWatchActionFn(value) {
+                $element.prop(propName, value && sanitizer(value));
```
The `value &&` check is redundant. `.prop(propName, sanitizer(value))` should be enough (and gives a chance to sanitizers to decide whether a falsy value is safe or not.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12111,620 @@ describe('$compile', function() {
+      expect($rootScope.name).toBe('Misko3');
+    }));
+
+    it("should use angular.element(x).on() API to add listener", inject(function($compile, $rootScope) {
```
`...to add listener` --> `...to add listener, so that Jason can hook his ngUpgrade hacks into it` 😛

jbedard replied:
:rofl:

gkalpak reviewed (commented):
A couple of thoughts:

- We also need to document how people can bind to case-sensitive properties using case-insensitive attributes (e.g. `ng-prop-inner_h_t_m_l`) :grin:

- It would be interesting to see how the compiler performance is affected by the changes.

Narretz commented:
Wrt compiler performance - I think we should put ngProp and ngOn behind a flag, similar to css class / comment directives. This way, there's no hit if you don't need it.

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,90 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+      'object|codebase',  'object|data',
+      'script|src',
+    ]);
+  })();
```
It hides the `registerContext` helper method, and I guess the main reason is to isolate the block of code (mostly) copied from Angular.

gkalpak replied:
Hide from who? We don't this IIRC.

jbedard replied:
Hide from you!

I guess I just like keeping the stuff copied from Angular in one place. Prevents people from using it or moving the helper method elsewhere etc.

I can drop the IIFE if people don't like it though...

gkalpak replied:
You keep talking about this `registerContext` helepr method, but I don't see it 😛

jbedard replied:
Perfect, the IIFE is hiding it well then 😛

petebacondarwin replied:
Does the IIFE help the closure compiler to reduce the code size?

jbedard replied:
As mentioned in a different comment... the IIFE probably adds 5-10 bytes (`"function(){}()".length` + gzipping). The closure compiler does not remove it like I was hoping, I guess because it would cause some vars/functions to be exposed a level higher.

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,90 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+      'input|src',
+      'ins|cite',
+      'q|cite',
+      'script|src',
```
Wow, not sure how I did that one :|

gkalpak replied:
Instead of `Ctrl+X, Ctrl+V` you do `Ctrl+C, Ctrl+V` (or `Cmd` or whatever you guys do on Macs 😛).

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,90 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * @param {string} ctx the context type
+   */
+  this.addPropertyContext = function(elementName, propertyName, ctx) {
+    var key = (elementName.toLowerCase() + '|' + propertyName.toLowerCase());
```
I think Angular may have done that for a reason?

https://github.com/angular/angular/blob/6.0.6/packages/compiler/src/schema/dom_security_schema.ts#L29

gkalpak replied:
Interesting. OK then :smiley:
(The only "native" properties that are affected are `formAction`, `innerHTML` and `outerHTML`. The last two are weird as custom element property names anyway.)

petebacondarwin replied:
So in the [commit for that line of code](https://github.com/angular/angular/commit/040b101842b2bdb637caff91a2c547a4317e5100):

> Comparison of (potentially) dangerous property names is done case insensitive,
to avoid issues like formAction vs formaction.

I guess they are saying that in "some" browsers, "some" properties may be assignable using different casing?

jbedard commented on docs/content/error/$compile/ctxoverride.ngdoc:
````diff
@@ -0,0 +1,13 @@
+```
+$compileProvider.addPropertyContext("my-element", "src", "mediaUrl");
+$compileProvider.addPropertyContext("my-element", "src", "resourceUrl");    //throws
+```
````
Which docs specifically?

gkalpak replied:
The `addPropertyContext` docs :grin:

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,90 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * @name $compileProvider#addPropertyContext
+   * @description
+   *
+   * Defines the security context for HTML properties bound by ng-prop-*
```
Looks like we normally say "HTML/SVG"?

gkalpak replied:
I don't know. Maybe :grin:
But `ngProp` should also work on SVG elements (which are first-class citizen in AngularJS 🤓).

petebacondarwin replied:
I don't think this is HTML at all, it should be "DOM properties" no?

jbedard commented on src/ng/compile.js:
```diff
@@ -2245,43 +2337,58 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             }
 
             nName = directiveNormalize(name.toLowerCase());
             attrsMap[nName] = name;
```
Is this right for ng-on/prop? I don't think we'd want to "map" `ng-on-foo` to the `foo` attribute... Would we want to map it at all? To `ngAttrName` (still containing ng-on/prop prefix) instead of `name`?

gkalpak replied:
It is fine imo. We should indeed not reflect the property/event as attribute, but keeping track of the `ngProp/On` prefixed attribute sounds reasonable and consistent (since the attribute is there).

jbedard replied:
I think your comment is true regarding the `attrs[ngAttrName] = value;` a few lines down. But what about the "mapping" on this line?

This line creates the `ngPropFoo => foo` mapping for https://docs.angularjs.org/api/ng/type/$compile.directive.Attributes#$attr ...

gkalpak replied:
Oh, you are right. I don't think we want to map `foo` to `ngPropFoo`. 
(Mapping `ng-prop-foo` to `ngPropFoo` sounds right (for consistency), although it would probably have no effect.)

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,90 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * 
+   * @param {string} elementName the element name or '*' to match any element
+   * @param {string} propertyName the property name
+   * @param {string} ctx the context type
```
The description of this parameter could be a bit more descriptive. This is what is on the `$sce.trustAs` method for a similar parameter:

> The context in which this value is safe for use, e.g. `$sce.URL`,
> `$sce.RESOURCE_URL`, `$sce.HTML`, `$sce.JS` or `$sce.CSS`.

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,90 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+   * - STYLE => CSS
+   * - various URL => MEDIA_URL
+   */
+  (function setupPropertyContexts() {
```
Tempting to name this function `registerNativePropertyContexts` or something similar?

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -2245,43 +2337,58 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+
+            if (isNgProp) {
+              attrs[ngAttrName] = value;
+              addPropertyDirective(node, directives, ngAttrName, name);
```
Why not pass `value` in here since we know it and it must be a static template string? That way you don't need to add the value to the `attrs` map?

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -3350,9 +3456,56 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+      directives.push({
+        priority: 100,
+        compile: function ngPropCompileFn(_, attr) {
+          var fn = $parse(attr[attrName]);
```
We could receive this value as a parameter to the `addPropertyDirective` and avoid adding it to the `attrs` map.

jbedard replied:
Note that I wanted to read the attribute here at compile time so it is equivalent to attribute interpolation and `$parse`ing of `ng-event` and `ng-on-*` which all read the attribute at compile time (although interpolation then reads it *again* at link time :|).

petebacondarwin commented on src/ng/directive/ngEventDirs.js:
```diff
@@ -51,32 +51,36 @@ forEach(
   }
 );
 
+function createEventDirective($parse, $rootScope, attrName, eventName, forceAsync) {
```
We could have a `value` param on this function which is used rather than `attr[attrName]` if it is defined.
That way you can avoid adding the value to the `attrs` map in the case of `ng-on-`, no?

jbedard replied:
I wanted to avoid making changes to the existing event directives, and I thought making `ng-on-*` the same would also be good. In all those cases the `attr[attrName]` is not read until compile time...

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -2245,43 +2337,58 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+              addPropertyDirective(node, directives, ngAttrName, name);
+            } else if (isNgEvent) {
+              attrs[ngAttrName] = value;
+              addEventDirective(directives, ngAttrName, name);
```
We could pass the `value` here as an extra param, and this function would know to use that rather than reading off the `attrs` map.

petebacondarwin reviewed (changes requested):
Maybe I am missing something but I feel really uncomfortable about re-using the `$attr` stuff for these property directives. I believe that a bit of refactoring would allow us to avoid putting these values on that map and give us tighter control over how these special directives are handled.

gkalpak commented:
Why? What is the problem with having attributes reflected on `$attrs`? Why is it OK for `$attrs.ngClick` to be present, but not for `$attrs.ngOnFooEvent`?

jbedard commented:
Similar to what @gkalpak said, I think `ngClick` and `ngOnClick` should be equivalent other then the extra "on" in the attribute name. Similar for properties... `ngPropDisabled` and `ngDisabled` should be equivalent, except for interpolation vs expression (and the fact that `ngDisabled` does some to-boolean casting).

I think this means that `$attrs` will contain `ngOnClick` and `ngPropDisabled`, but will not contain `click` or `disabled` or anything like that...

petebacondarwin commented:
Well I never! `ngClick` does get added to the attrs. That is interesting...

OK so I relent on my campaign to remove it. I'll take a look at the security stuff and tests later today.

jbedard commented:
Added a few more fixups including a fix for the `attr.$attr` mappings (https://github.com/angular/angular.js/pull/16614/commits/4995d0e8d46fadd0692631bd704f8c9c6f54e1b2) along with some tests that failed before it (the ng-attr test never failed, but seemed good to add at the same time).

gkalpak commented on src/ng/compile.js:
```diff
@@ -2245,43 +2336,62 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-            if (isNgAttr || !attrs.hasOwnProperty(nName)) {
+            if (isNgProp || isNgEvent) {
+              attrs[nName] = value;
+              attrsMap[nName] = attr.name;
```
For reference, this is the difference between how `ngAttr` and `ngProp` attributes get mapped (and it looks reasonable to me):

```js
// HTML: x-ng-attr-my-form_action
attrsMap['myFormaction'] = 'my-formAction'

// HTML: x-ng-prop-my-form_action
attrsMap['ngPropMyForm_action'] = 'x-ng-prop-my-form_action'
```

jbedard replied:
Should the `_`s not behave the same?

I'll try to add some tests with `_`s for both `ngAttr` and `ngProp`

jbedard replied:
I've added a commit https://github.com/angular/angular.js/pull/16614/commits/c032255cbaa3ca9b748f4d4b4dc6fe70bb5286c8 (or two https://github.com/angular/angular.js/pull/16614/commits/7835fc36c65178ee6985f797e898c449aee35e81) to test this.

The result is a bit different then you assumed for the ngProp case:
```
// HTML: x-ng-attr-my-form_action
attrsMap['myFormaction'] = 'my-formAction'

// HTML: x-ng-prop-my-form_action
attrsMap['ngPropMyFormAction'] = 'x-ng-prop-my-form_action'
```

Notice the capital A for the ng-prop one... I think that is actually more correct then the ng-attr one. This is because we invoke `directiveNormalize` in two different ways:
* https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L2258 - on the raw attribute name
* https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L2274 - on the no-prefix attribute name, but in all lower case. This lower casing causes the mapping to be `myFormaction` instead of `myFormAction`

Is this a bug? Or a feature and we should be copying for `ng-prop`? :thinking:

jbedard replied:
Yeah... the `element.attr('...')` and the real attribute name are correct, it's only the [nName](https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L2274) variable which is what is used on `$attrs` that is wrong.

Should we fix it though? If anyone is depending on these bad names it would be hard to find+fix if we fixed the bug. However it would be nice to fix the bug...

jbedard replied:
Created https://github.com/angular/angular.js/issues/16624

petebacondarwin commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12141,626 @@ describe('$compile', function() {
+        });
+      });
+
+      it('should not have endless digests when given arrays in concatenable context', inject(function($compile, $rootScope) {
```
Didn't we agree that ng-prop values were not concatenable?

jbedard replied:
I thought this one might be worth it still, although could probably be renamed. I think I mainly wanted to test that using an object literal doesn't cause an `infdig`.

petebacondarwin commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12141,626 @@ describe('$compile', function() {
+      }));
+
+      it('should accept trusted values for non-whitelisted values', inject(function($rootScope, $compile, $sce) {
+        $rootScope.testUrl = $sce.trustAsUrl('javascript:foo()'); // `javascript` is not whitelisted
```
Similar to `img[src]` we should check the different trust levels. E.g. ResourceUrl and Url should pass, but in this case MediaUrl should fail, right?

jbedard replied:
I'm really not sure which should pass/fail. I'll need some help with those tests (which are still a TODO).

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -3350,9 +3459,56 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+          var fn = $parse(attr[attrName]);
+          return {
+            pre: function ngPropPreLinkFn(scope, $element) {
+              scope.$watch(fn, function propertyWatchActionFn(value) {
```
Do we need to look at the implementation of `ngBindHtml` here for the watcher, where the inner value is what is watched rather than the trusted object; and also consider calls to `$$bindingInfo` and friends.

See https://github.com/angular/angular.js/blob/74726b4f7934720c8106dfb7212b868ad5efc111/src/ng/directive/ngBind.js#L189-L206

jbedard replied:
I've essentially copied `ngBindHtml`: https://github.com/angular/angular.js/pull/16614/commits/f547b0da82724b1d9d2f3ec38f8c7332bd4f1933

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -3350,9 +3459,56 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+          var fn = $parse(attr[attrName]);
+          return {
+            pre: function ngPropPreLinkFn(scope, $element) {
+              scope.$watch(fn, function propertyWatchActionFn(value) {
```
Also, although the SCE stuff should take care of sanitising and protecting things like `innerHTML` I think it would be sensible to have a few tests here either to check that the call to `$sce,getTrusted` is made or that `innerHTML` is sanitised correctly.

jbedard replied:
I still need to write a lot of sce related tests

jbedard replied:
I think https://github.com/angular/angular.js/pull/16614/commits/f547b0da82724b1d9d2f3ec38f8c7332bd4f1933 covers the specific cases you mentioned?

jbedard commented on src/ng/compile.js:
```diff
@@ -3350,9 +3459,56 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+
+      var sanitizer = identity;
+      // Sanitize img[srcset] + source[srcset] values.
+      if ((nodeName === 'img' || nodeName === 'source') && propName === 'srcset') {
```
Note that I added the `<source>` support because Angular has it. I should add tests for both attr+prop...

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12141,626 @@ describe('$compile', function() {
+
+    it('should work with different prefixes', inject(function($rootScope, $compile) {
+      $rootScope.name = 'Misko';
+      element = $compile('<span ng:prop:test="name" ng-Prop-test2="name" ng_Prop_test3="name"></span>')($rootScope);
```
For some reason we support mixed case ng-attr/prop prefixes...

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12026,6 +12026,36 @@ describe('$compile', function() {
+        expect(attrs.$attr.ngAttrSuperTitle).toBeUndefined();
+
+        expect(attrs.myCameltitle).toBe('56');
+        expect(attrs.$attr.myCameltitle).toBe('my-camelTitle');
```
This is what is being discussed above: https://github.com/angular/angular.js/pull/16614#discussion_r200251353

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12141,626 @@ describe('$compile', function() {
+
+    it('should work with the "href" property', inject(function($rootScope, $compile) {
+      $rootScope.value = 'test';
+      element = $compile('<a ng-prop-href="\'test/\' + value"></a>')($rootScope);
```
No. Just like `ng-attr-href` won't (I'm pretty sure?).

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12141,626 @@ describe('$compile', function() {
+      $rootScope.asdf = 123;
+      $rootScope.$digest();
+      expect(element.prop('asdf')).toBe(123);
+      expect(element.attr('asdf')).toBe('foo');
```
Depends what order the browser lists the attributes, which might depend on the order in the DOM.

jbedard commented on src/ng/compile.js:
```diff
@@ -1587,6 +1587,89 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+    registerContext(SCE_CONTEXTS.MEDIA_URL, [
+      'audio|src',
+      'img|src',    'img|srcset',
+      'source|src', 'source|srcset',
```
Yeah I guess that's true. But maybe leave them just so people can't override them to a different context type?

jbedard commented on src/ng/compile.js:
```diff
@@ -1767,52 +1901,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-          }
-          this[key] = value = result;
+        // Sanitize img[srcset] + source[srcset] values.
+        if ((nodeName === 'img' || nodeName === 'source') && key === 'srcset') {
```
Correct. I added it since Angular sanitized it (at least the property). Should I undo this?

I have a feeling we didn't have this because the `source[srcset]` didn't exist when this code was written? :older_man:

petebacondarwin replied:
That is probably true but I think it would still count as a BC?

jbedard replied:
Yes it would. Shall we just leave this as is? The `source[srcset]` attribute will have no sce like it always has, but the property will...

jbedard replied:
I reverted this to avoid changing to attribute sanitation, mainly because it is a little unrelated and would be a BC.

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12141,626 @@ describe('$compile', function() {
+          'Please use the ng- versions (such as ng-click or ng-on-click instead of ng-prop-onclick) instead.');
+    }));
+
+    it('should process property binding in pre-linking phase at priority 100', function() {
```
This description was copied from an `ng-attr-*` test, but doesn't describe the logic of `ng-prop-*` because we only assign the prop in the `$watch`er. This is unlike `ng-attr-*` that has special logic to assign the attribute asap. See discussion: https://github.com/angular/angular.js/pull/16614#discussion_r198717403

jbedard replied:
For now I'll just update the description of this test.

jbedard commented:
I've added the `sce.valueOf` watch logic and copied tests from `ngBindHtml`. WDYT about copying tests like this? It's basically an exact copy just replacing `ng-bind-html` with `ng-prop-inner_h_t_m_l`...

https://github.com/angular/angular.js/pull/16614/commits/f547b0da82724b1d9d2f3ec38f8c7332bd4f1933

jbedard commented on src/ng/compile.js:
```diff
@@ -3370,7 +3526,7 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
         throw $compileMinErr('nodomevents',
             'Interpolations for HTML DOM event attributes are disallowed.  Please use the ' +
-                'ng- versions (such as ng-click instead of onclick) instead.');
+                'ng- versions (such as ng-click or ng-on-click instead of onclick) instead.');
```
I've updated this, as well as the interpolate one (by accident). Ok with updating the interpolate as well?

jbedard commented:
I've added the `source[srcset]` tests: https://github.com/angular/angular.js/pull/16614/commits/d1632437590510be3597bac87d6300875fe3e67b

jbedard commented:
I've added tests for `iframe[src]`, `base[href]`, `form[action]`, `link[href]`. Note that I changed `*|formAction` and `form|action` to be `RESOURCE_URL` like the attribute version of `form|action`. This is another difference from Angular where they are `URL` and was noted in a comment.

a93ee19e2251545ed2b5d2977ddc01ddcc7d98d7

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -3350,9 +3460,60 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+      if (EVENT_HANDLER_ATTR_REGEXP.test(propName)) {
+        throw $compileMinErr('nodomevents',
+            'Property bindings for HTML DOM event properties are disallowed.  Please use the ' +
+                'ng- versions (such as ng-click or ng-on-click rather than ng-prop-onclick) instead.');
```
Tiny NIT - this error message is rather long. Perhaps we should just update the error doc with the workaround info?

petebacondarwin commented on src/ng/compile.js:
```diff
@@ -3370,7 +3531,7 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
         throw $compileMinErr('nodomevents',
             'Interpolations for HTML DOM event attributes are disallowed.  Please use the ' +
-                'ng- versions (such as ng-click instead of onclick) instead.');
+                'ng- versions (such as ng-click or ng-on-click rather than onclick) instead.');
```
Same here.

jbedard replied:
These are definitely long. I guess if I'm changing them already then there's no harm also making them shorter? Think we should do that in this PR?

petebacondarwin replied:
If you feel you have capacity to do that. (And the saved bytes would make up for the extra sanitizer code that we have landed here :-) )

jbedard replied:
I think I can do it along with the other changes 👍 

I'll adjust the docs to ensure everything in this message is in there, and I think this message can be cut down to just the first sentence. Sound ok?

jbedard replied:
Done: a3b1f41f0350880a4b9e7eefaa13e67d0bf45776

petebacondarwin commented on test/ng/compileSpec.js:
```diff
@@ -12962,6 +12962,27 @@ describe('$compile', function() {
+      it('should NOT set style for untrusted values', inject(function($rootScope, $compile) {
+        element = $compile('<div ng-prop-style="style"></div>')($rootScope);
+        $rootScope.style = 'background: red';
+        expect(function() { $rootScope.$digest(); }).toThrow();
```
Should we be more strict on what is thrown here? E.g. `.toThrowMinErr('$sce', 'insecurl', ...)`?

jbedard replied:
Done (b3563c58db80721795542ec8d9e27ec0936e1d42), but note it is a `unsafe` error in these cases.

petebacondarwin commented on test/ng/compileSpec.js:
```diff
@@ -12719,6 +12719,127 @@ describe('$compile', function() {
+        it('should NOT set html for untrusted values', inject(function($rootScope, $compile) {
+          element = $compile('<div ng-prop-inner_h_t_m_l="html"></div>')($rootScope);
+          $rootScope.html = '<div onclick="">hello</div>';
+          expect(function() { $rootScope.$digest(); }).toThrow();
```
Should we check for a specific error here?

petebacondarwin reviewed (approved):
LGTM - a couple of tiny suggestions but otherwise good to land, IMO.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+          $rootScope.$digest();
+          expect(element.prop('srcset')).toBe('http://example.com/');
+
+          // Now set it to undefined
```
Undefined you say... 🤔

jbedard replied:
The problems with copy pasta...

jbedard replied:
Will also update the test description because actually it should error when set to undefined, because the property converts that to the `"undefined"` string value.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+          expect(element.prop('srcset')).toEqual('http://example.com');
+        }));
+
+        it('does not work with trusted values', inject(function($rootScope, $compile, $sce) {
```
_does_ --> _should_
_not_ --> _NOT_? 😛

jbedard replied:
Again this was copied from the attr version. Shall I change both?

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+            element = $compile('<' + srcsetElement + ' ng-prop-srcset="testUrl + \',\' + testUrl"></' + srcsetElement + '>')($rootScope);
+            $rootScope.testUrl = 'javascript:yay';
+            $rootScope.$apply();
+            expect(element.prop('srcset')).toEqual('someSanitizedUrl ,someSanitizedUrl');
```
Why the weird "whitespacing"? Who does that?

jbedard replied:
Typo I assume...

jbedard replied:
Actually... I have no idea. The attribute version has the same thing. Something to do with the special `srcset` sanitizing?

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+      }));
+
+
+      it('should pass through $sce.trustAs() values in action property', inject(function($compile, $rootScope, $sce) {
```
`trustAs()` --> `trustAsResourceUrl()`

jbedard replied:
Again... make it different then the attr version?

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+            '  URL: https://evil.example.org/css.css');
+      }));
+
+      it('should accept valid RESOURCE_URLs', inject(function($compile, $rootScope, $sce) {
```
NIT: It is not about "valid"/"invalid". It's more about "trusted"/"untrusted".

jbedard replied:
Attr version... (I definitely agree with your comment, but should it differ?)

gkalpak replied:
Generally, when I find something wrong I try to fix it (or at least not replicate it 😛).
But it is not critical or anything.

jbedard replied:
Sounds good. Shall I also fix the attr version?

jbedard replied:
Actually... when you say "trusted" does that mean an instanced of the `$sce.trustAs*()` object? Because this tests both that object as well as other "valid" things like relative paths. But that "valid" word can also be swapped for "trusted"... preference?

gkalpak replied:
By "trusted" I mean anything that is considered safe (either by being wrapped in `$sce.trustAs()` or being included in a whitelist or having some other property (e.g. being a relative URL) that makes it safe in a particular context).

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+  describe('addPropertySecurityContext', function() {
+    function testProvider(provider) {
+      module(provider);
+      inject(function($compile) { /* done! */});
```
Missing whitespace before closing `}` 😇

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+        $compileProvider.addPropertySecurityContext('div', 'title', 'mediaUrl');
+      });
+    });
+  });
```
We also need tests that verify the effect of `addPropertySecurityContext()` :grin:

jbedard replied:
See fae50665e

gkalpak reviewed (changes requested):
We're getting there! I'm so excited about this :tada:

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12026,6 +12026,37 @@ describe('$compile', function() {
+        expect(attrs.ngAttrSuperTitle).toBeUndefined();
+        expect(attrs.$attr.ngAttrSuperTitle).toBeUndefined();
+
+        //Note the casing is incorrect: https://github.com/angular/angular.js/issues/16624
```
It's definitely not a habbit of mine 😛

jbedard commented on src/ng/compile.js:
```diff
@@ -3350,9 +3460,60 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+            return $sce.valueOf(val);
+          });
+
+          return function ngPropPreLinkFn(scope, $element) {
```
I guess I'll switch it back to pre, although I don't think it really matters (because all we do is add a watcher, there is no assigning of the initial value like `ngAttr*`).

gkalpak replied:
Maybe there should be assigning too 🤔

jbedard replied:
See https://github.com/angular/angular.js/pull/16614#discussion_r198717403

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+      $rootScope.$digest();
+      expect(element.prop('disabled')).toBe(false);
+      $rootScope.isDisabled = true;
+      $rootScope.$digest();
```
But then it has to be `$parse`ed!

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12142,890 @@ describe('$compile', function() {
+          };
+        });
+      });
+      module(function() {
```
Because I copied the attr version which does this 🤷‍♂️

jbedard commented on docs/content/error/$compile/ctxoverride.ngdoc:
```diff
@@ -0,0 +1,13 @@
+@fullName Context Override
+@description
+
+This error occurs when the security context for a property is defined multiple times under different security contexts.
```
Would this just be something such as: "See https://docs.angularjs.org/api/ng/provider/$compileProvider#addPropertySecurityContext"?

gkalpak replied:
Not sure what you are asking 😕. If you are asking about the text, it can be whatever you want (as long as there is a link to `addPropertySecurityContext`). 

(If you are asking about how to create the link, it will something like `{@link ng.$compilerProvider#addPropertySecurityContext addPropertySecurityContext()}`.)

jbedard replied:
Was mainly wondering about the `@link` (and couldn't find an example in my 30s attempt) :+1:

jbedard commented on src/ng/compile.js:
```diff
@@ -3317,11 +3428,10 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
 
 
-    function getTrustedContext(node, attrNormalizedName) {
+    function getTrustedAttrContext(tag, attrNormalizedName) {
```
Actually this function previously had `var tag = nodeName_(node)`, so changing this will increase the diff. Still want to change it for the consistency?

gkalpak replied:
I'd take larger diff over consistency any day. (But I admit that's a personal preference :grin:)
Up to you.

jbedard replied:
Done: d1264cde1

jbedard commented on src/ng/compile.js:
```diff
@@ -1767,52 +1902,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-          }
-          this[key] = value = result;
+        // Sanitize img[srcset] + source[srcset] values.
+        if (nodeName === 'img' && key === 'srcset') {
```
Note that I originally changed this to also include `source` for consistency with Angular + ng-prop, but this was revert to avoid breaking changes. But I forgot to revert the comment, so I'll be addressing this comment by reverting the comment back to how it is in master...

jbedard commented:
@gkalpak I've addressed most of your comments (3bf86d2c525ef17fdb1cdd38d4251f35fc7731f9).

I think we need to decide the [props should be preassigned](https://github.com/angular/angular.js/pull/16614#discussion_r198717403) like attributes.

I've added TODOs to the main PR description regarding the tests you wanted to add (not sure if I'll get to that tonight...).

jbedard commented on test/ng/compileSpec.js:
```diff
@@ -12111,6 +12138,883 @@ describe('$compile', function() {
+      });
+    });
+
+    it('should setting the same property/element to the same value', function() {
```
should *allow*

Narretz commented on docs/content/error/$compile/ctxoverride.ngdoc:
```diff
@@ -0,0 +1,13 @@
+@fullName DOM Property Security Context Overrid
+@description
+
+This error occurs when the security context for a property is defined via {@link ng.$compilerProvider#addPropertySecurityContext addPropertySecurityContext()} multiple times under different security contexts.
```
This link is incorrect

gkalpak replied:
That's what you get for blindly copy-pasting my suggestions 😛

Narretz replied:
Still not fixed :<

jbedard commented:
I've moved the ngProp+ngOn specs to their own files and reverted a couple of the compileSpec changes so the full diff is a bit less (7157abe). I think I prefer this, but let me know if you think any file names or descriptions should change.

jbedard commented:
I might be forgetting everything we discussed last week but... for attributes it actually only pre-assigns  [on the `$attrs` object](https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L3400) and not on the actual dom which [occurs on `$watch`](https://github.com/angular/angular.js/blob/v1.7.2/src/ng/compile.js#L3414). This is shown [in some tests](https://github.com/angular/angular.js/blob/v1.7.2/test/ng/compileSpec.js#L12020-L12026)).

From the perspective of web-components it might make sense to align the dom attribute+property set-times.

From the perspective of AngularJS components it might make sense to align the dom property set-times with `$attrs` set times. Or is that comparing two unrelated things?

I keep flip-flopping on this one...

gkalpak commented:
> So a consumer could get the property value at an earlier time by evaluating the expression by accessing it from `$attrs`. If this is correct I think this is an okay situation.

It doesn't sound that OK to me 😛
If we have the value, then I see no reason to postpone setting it. The whole purpose of putting the attr value on `$attrs` is so that it can be accessed by other directive's linking functions (especially for isolate scope initialization):

https://github.com/angular/angular.js/blob/6706353a71e3c11c56c0b19be03600dac57aafe1/src/ng/compile.js#L3397-L3399

Attributes have both `$attrs` and `getAttribute()`, but with props there is no alternative.

And accessing the DOM should be much of an issue for most properties I guess (especially for custom element properties); i.e. no reflow should be triggered.

gkalpak commented on docs/content/error/$compile/ctxoverride.ngdoc:
```diff
@@ -0,0 +1,13 @@
+@ngdoc error
+@name $compile:ctxoverride
+@fullName DOM Property Security Context Overrid
```
_Overrid_ --> _Override_ (?)

jbedard commented:
@gkalpak so you'd prefer aligning `$e.prop(x)`/`$attrs[x]` instead of `$e.prop(x)`/`$e.attr(x)`?

As an "angularjs api" I agree with aligning the primary way of retrieving these attributes/props. But if these props/attributes were being passed to web components it would be weird that props and attributes get initialized at different times...

gkalpak commented:
> But if these props/attributes were being passed to web components it would be weird that props and attributes get initialized at different times...

Kind of true, but theoretically this shouldn't make a difference for custom elements.

Generally, from a custom element's standpoint, setting properties and attributes via `ngProp`/`ngAttr` happens asynchronously. So, I wouldn't be to worried about the custom elements themselves.
From an AngularJS standpoint, having the properties earlier will possibly make it easier for other directives that care and attributes will also be available at the same time via `$attrs`.

(But I don't feel too strongly about it.)

jbedard commented:
> From an AngularJS standpoint, having the properties earlier will possibly make it easier for other directives that care

I guess that might convince me (and maybe that's what we concluded in the last meeting?). So for now I'll keep 13bab70bfa3ad733fafc050e1431e32f9b7a6021...

jbedard commented on test/ng/ngPropSpec.js:
```diff
@@ -0,0 +1,775 @@
+
+  });
+
+  describe('*[style]', function() {
```
I think these tests are actually all invalid because you can't assign to `HTMLElement.style`. The attribute version can be assigned as a plain string, but the property is of type `CSSStyleDeclaration` and can't be assigned.

The tests are failing in some browsers which seem to throw when assigning to `HTMLElement.style`.

Should we have any special handling for non-assignable properties like this? Should we drop the `*|style => $sce.CSS` declaration?

jbedard replied:
Although some browsers seem to accept string assignments, so maybe these tests just need an `if (canAssignToStyleProp)` around them...

jbedard replied:
Anyone have suggestions for how to do the `canAssignToStyleProp`? Only thing I can think of is:
```
  function canAssingStyleProp() {
    try {
      document.createElement("div").style = "background: red";
      return true;
    } catch (e) {
      return false;
    }
  }
```
... however I'd rather not depend on something like that in tests.

gkalpak replied:
Why not? (BTW, there are browsers (e.g. IE11) that don't throw an error but ignore the assignment.)

jbedard replied:
> Why not?

Because I don't trust my feature-test in tests, unless we tested them. But if we can't think of anything else I'll do it. Might use the `msie` variable for the IE detection though...

gkalpak replied:
Do you know which browsers are problematic?

jbedard replied:
I think https://travis-ci.org/angular/angular.js/jobs/406066641#L771 has a good summary, but it's hard to tell which failures are for which browser. Is there a way of seeing that? Logs somewhere?

But you can see the following errors:
* Expected '0px' to equal '10px'.
* TypeError: Attempted to assign to readonly property. (line 7653)
* Expected '' to equal '10px'

(there's also an undefined var in there which I'll push a fix for in a sec...)

gkalpak replied:
Weird that the browser reporting the error is not shown. Maybe our reporter is outdated?
I don't know of any other way to find it. But I would be fine adding a `canAssingStyleProp()` check fwiw :grin:

jbedard replied:
I've added 08898499df2c393e724cacf7b26f887e510ba911, but I'm taking a guess at which browser does `''` vs `'0px'` (and I'm probably wrong now that I think about it more) and put `// Support: ??` for the case where it throws.

jbedard replied:
My guess was correct!

Just one odd IE9 error now... https://travis-ci.org/angular/angular.js/jobs/406368530#L687

jbedard commented on test/ng/ngPropSpec.js:
```diff
@@ -0,0 +1,775 @@
+  });
+
+  ['img', 'source'].forEach(function(srcsetElement) {
+    describe(srcsetElement + '[srcset] sanitization', function() {
```
These seem to fail in Edge for `source[srcset]` because the `srcset` property remains `undefined`. Any ideas why assigning this property in Edge would be rejected and remain `undefined`?

jbedard replied:
See https://github.com/angular/angular.js/commit/9de0115711d0963c85eef2faa81242234a5c5f15?w=1

jbedard commented on test/ng/ngPropSpec.js:
```diff
@@ -0,0 +1,775 @@
+      var element = $compile('<div ng-prop-style="style"></div>')($rootScope);
+      $rootScope.style = $sce.trustAsCss('background: red');
+      $rootScope.$digest();
+      expect(lowercase(element.css('background-color'))).toEqual('red');
```
A lot of browsers (which allow assigning to `HTMLElement.style`) seem to return `rgb(255,0,0)` instead of `red` here. I'm going to change this to a property that won't have normalization issues like that.

jbedard replied:
See 125c7b138374409ca96fe51bd125219a518d8c1f

jbedard commented:
I've disabled the `source[srcset]` tests in IE9 (https://github.com/angular/angular.js/commit/9de0115711d0963c85eef2faa81242234a5c5f15?w=1) due to the failures you can see in https://travis-ci.org/angular/angular.js/builds/405950267

jbedard commented:
I've added the remaining tests from the TODO list: 14119a9e3

Fixed the last failing test in IE9: 0beb9bab03cdb3cc742b16d6199dd49f2389f4e8

Added some comments, including a reference to the issue causing #16624: 6ec8a682088250a906d3101016c6a22f41a5f1eb, d30c214e9056deb06367d8390973385ad373b055

jbedard commented:
I think all the feedback has been addressed. Let me know if there is anything else. I'll rebase+squash once this has had another look...

jbedard commented on src/ng/compile.js:
```diff
@@ -2245,43 +2338,66 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-            ngAttrName = directiveNormalize(name);
-            isNgAttr = NG_ATTR_BINDING.test(ngAttrName);
-            if (isNgAttr) {
+            name = name.toLowerCase();
```
This is an interesting one. Note that previously `name` did not have the `.toLowerCase()` assigned to it, we only did:
```
nName = directiveNormalize(name.toLowerCase());
```
Also... `attr.name` should always be lower case anyway, since browsers normalize attribute names to all lower case. However without this line one test failed in IE9: https://travis-ci.org/angular/angular.js/jobs/406368530#L687. An alternative would be making `EVENT_HANDLER_ATTR_REGEXP` case-insensitive, which I'm tempted to do either way...

Another weird one... if the `.toLowerCase()` is done above as `name = attr.name.toLowerCase()` then a lot of animation tests fail. (I seem to recall something about SVG attributes being different? Didn't check if the tests used SVG though...).

jbedard replied:
TODO: move this into the event/prop specific code to avoid unnecessary changes

jbedard commented on src/ng/compile.js:
```diff
@@ -2244,43 +2337,66 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+              // Normalize the non-prefixed name
               name = name.replace(PREFIX_REGEXP, '')
-                .substr(8).replace(/_(.)/g, function(match, letter) {
+                .toLowerCase()
```
Note that this was the `toLowerCase()` required for `name === "click"` when doing `ng-on-CLICK` in IE9. This may effect `ng-attr-FOO` in IE9 as well...

Narretz commented on test/ng/ngPropSpec.js:
```diff
@@ -0,0 +1,836 @@
+  describe('*[style]', function() {
+    // Support: ??
+    // Some browsers throw when assignging to HTMLElement.style
+    function canAssingStyleProp() {
```
typo: -> **canAssignStyleProp**

Narretz commented on test/ng/ngPropSpec.js:
```diff
@@ -0,0 +1,836 @@
+  });
+
+  describe('*[style]', function() {
+    // Support: ??
```
Don't we know which browser is this? IE9?

jbedard replied:
Where I was looking all the browsers test output as randomly output mixed together so I couldn't tell. Are there per-browser logs somewhere?

See https://travis-ci.org/angular/angular.js/jobs/406066641#L750

jbedard replied:
Although I would guess it is IE9 because 2 tests failed in IE9 and that error ("Attempted to assign to readonly property") occurred 2 times...

gkalpak reviewed (approved):
:100: x :tada:

gkalpak commented:
There is a typo in the commit message:

> [...] events could be bound using such as `ng-click` [...]

##
The following also sounds a little confusing (since there are no capital letters in the attribute to put the underscore before :grin:):

> For properties or events using mixed case underscores can be used before capital letters.

jbedard commented:
I've broken up the commit message into <100 char lines. Fixed the events sentance.

> The following also sounds a little confusing (since there are no capital letters in the attribute to put the underscore before 😁):

I added a comma in there. Is that clearer?

Thanks for reviewing the commit message. Let me know if you find any other issues or things that could be clearer or added.

gkalpak commented:
I also noticed that some double-quotes in the commit message are fancier than others (`”` vs `"`). It's good to have non-fancy quotes everywhere.

I would expand a bit on the "binding to camel case" thing. Maybe something like:

> Since HTML attributes are case-insensitive, you need to convert property and event names to snake_case for `ng-prop-*` and `ng-on-*`. For example, to bind to the `fooBar` property use `ng-prop-foo_bar`...

jbedard commented:
I've updated the message again

gkalpak commented:
I still see at least on fancy quote: `“bar"` :grin:
Other than that looks great 💯
:shipit:

jbedard commented:
Fixed the quotes again
